### PR TITLE
Add Decimal type with 14 built-in operations

### LINKED ISSUE #333: Decimal type for exact decimal arithmetic
Exact decimal arithmetic for financial and precision-sensitive applications. Software implementation in the runtime (WebAssembly has no native decimal floating-point).

Split from #62 (which originally covered Map, Set, and Decimal). Map and Set are being implemented in PRs #332 and a follow-up. Decimal is independent of the Map/JSON/HTTP critical path and can be scheduled separately.

**Spec reference:** Chapter 9, Section 9.7.2

### PR DESCRIPTION
**Pull request by aallan**
## Summary

- **Decimal type** ([#333](https://github.com/aallan/vera/issues/333)) — 14 host-import operations for exact decimal arithmetic, following the Map/Set opaque handle pattern
- **Operations**: `decimal_from_int`, `decimal_from_float`, `decimal_from_string`, `decimal_to_string`, `decimal_to_float`, `decimal_add`, `decimal_sub`, `decimal_mul`, `decimal_div`, `decimal_neg`, `decimal_compare`, `decimal_eq`, `decimal_round`, `decimal_abs`
- **Runtime**: Python uses `decimal.Decimal`; browser uses string-based arithmetic
- **Tests**: 17 new unit tests (7 checker + 10 codegen), conformance test `ch09_decimal` (59 programs)
- **Spec**: §9.7.2 updated with full documentation of all operations
- **Docs**: SKILL.md, ROADMAP, CHANGELOG, TESTING.md all updated
- **Version**: Bumped to v0.0.97
- **ROADMAP**: Ticked off #62 Map in the ASCII diagram

### Known limitations

- `Option<Decimal>` from `decimal_div`/`decimal_from_string` cannot be used with `match` or `option_unwrap_or` due to monomorphization gaps (Decimal is not recognized as a monomorphizable type parameter)
- `Ordering` from `decimal_compare` cannot be used with `match` for the same reason
- These limitations affect all new opaque types and are tracked as a broader codegen issue

Closes #333

## Test plan
- [x] All 21 pre-commit hooks pass
- [x] 2,853 tests pass
- [x] 59 conformance programs pass
- [x] 26 examples pass
- [x] mypy clean
- [ ] CI passes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a Decimal type for exact decimal arithmetic with 14 built-in operations and browser runtime support.

* **Documentation**
  * Spec, guides and READMEs updated with Decimal docs, examples and a limitation about using Option<Decimal>/Ordering with generic match/unwrapping.

* **Tests**
  * Conformance suite increased to 59 programs; new unit and end-to-end tests cover Decimal behaviour and edge cases.

* **Chores**
  * Project version bumped to 0.0.97.

### CODE REVIEW

File: ROADMAP.md [hunk 1]
```diff
@@ -75,7 +75,7 @@ Independent of the Tier 1 chain. Can be scheduled between Tier 1 items or in par
...
 - [#61](https://github.com/aallan/vera/issues/61) **Inference effect** — `effects(<Inference>)` in a signature means "this function calls an LLM, and you can mock it for testing." The feature that most differentiates Vera from Dafny as a verification target, and the one that positions it as the natural language for verified LLM orchestration.
-- [#333](https://github.com/aallan/vera/issues/333) **Decimal type** — exact decimal arithmetic for financial and precision-sensitive applications. Software implementation in the runtime (no native WASM decimal). Split from #62.
+- [#333](https://github.com/aallan/vera/issues/333) **Decimal type** — exact decimal arithmetic for financial and precision-sensitive applications. 14 built-in operations (construction, arithmetic, comparison, rounding, conversion) via host imports, following the Map/Set pattern. Software implementation in the runtime (Python `decimal.Decimal` / JS string-based decimal). Known limitation: `Option<Decimal>` from `decimal_div`/`decimal_from_string` and `Ordering` from `decimal_compare` cannot yet be used with `match` due to monomorphization gaps.
```
[the remark this answers is not in the document]
Comment (aallan) [reply ROADMAP.md]: @coderabbitai Fixed — added `option_unwrap_or` to the limitation description in ROADMAP.md line 78.

File: scripts/check_skill_examples.py [hunk 2]
```diff
@@ -26,117 +26,124 @@
...
-    496: ("FRAGMENT", "Array built-in examples, bare calls"),
+    497: ("FRAGMENT", "Array built-in examples, bare calls"),
 
     # Map operations — bare function calls
-    514: ("FRAGMENT", "Map built-in examples, bare calls"),
+    515: ("FRAGMENT", "Map built-in examples, bare calls"),
 
     # Handler syntax — pseudocode template
-    952: ("FRAGMENT", "Effect handler syntax template"),
+    974: ("FRAGMENT", "Effect handler syntax template"),
```
[the remark this answers is not in the document]
Comment (aallan) [reply scripts/check_skill_examples.py]: @coderabbitai Fixed — removed 3 duplicate ALLOWLIST keys (974, 1274, 1364). Each now appears only once.

File: spec/09-standard-library.md [hunk 3]
```diff
@@ -1510,11 +1510,55 @@ type ApiResponse = { @Json | has_field(@Json.0, "status") };
...
+  requires(true)
+  ensures(@Int.result == 1)
+  effects(pure)
+{
+  let @Decimal = decimal_add(decimal_from_int(100), decimal_from_int(3));
+  if decimal_eq(@Decimal.0, decimal_from_int(103)) then { 1 } else { 0 }
+}
+```
+
+**Known limitation:** `Option<Decimal>` values returned by `decimal_div` and `decimal_from_string` cannot currently be used with `option_unwrap_or` or `match` because the monomorphizer does not produce `option_unwrap_or$Decimal`. This affects all new opaque types and will be resolved in a future release.
```
[the remark this answers is not in the document]
Comment (aallan) [reply spec/09-standard-library.md]: @coderabbitai Fixed — added `Ordering` from `decimal_compare` to the spec limitation note.

File: tests/test_codegen.py [hunk 4]
```diff
@@ -8122,3 +8122,118 @@ def test_set_zero_value_element(self) -> None:
...
+        """neg(42) + 42 = 0."""
+        source = """
+public fn main(-> @Int)
+  requires(true) ensures(true) effects(pure)
+{
+  let @Decimal = decimal_add(decimal_neg(decimal_from_int(42)), decimal_from_int(42));
+  if decimal_eq(@Decimal.0, decimal_from_int(0)) then { 1 } else { 0 }
+}
+"""
+        assert _run(source) == 1
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen.py]: @coderabbitai Added `test_decimal_neg_zero` — verifies `decimal_neg(decimal_from_int(0))` equals `decimal_from_int(0)`. Also fixed browser runtime to canonicalize '-0' to '0'.

File: tests/test_checker.py [hunk 5]
```diff
@@ -1707,6 +1707,65 @@ def test_set_add_wrong_element_type(self) -> None:
...
+""", "expects")
+
+    def test_decimal_wrong_type(self) -> None:
+        """decimal_add with Int args produces error."""
+        _check_err("""
+private fn foo(@Unit -> @Decimal)
+  requires(true) ensures(true) effects(pure)
+{ decimal_add(1, 2) }
+""", "type")
+
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_checker.py]: @coderabbitai Added all missing tests: 3 happy-path (`decimal_from_float`, `decimal_sub`, `decimal_mul`) and 9 wrong-type tests (`decimal_from_float`, `decimal_sub`, `decimal_mul`, `decimal_neg`, `decimal_abs`, `decimal_round`, `decimal_compare`, `decimal_div`, `decimal_from_string`). Total: 25 Decimal checker tests.

**Comment by aallan** [comment]
@coderabbitai Addressed all 4 items from the latest review:

1. **Spec §9.1 / §9.7 consistency**: Moved Decimal out of the 'Future types' list and into its own bullet. Renamed §9.7 from 'Built-in Types (Future)' to 'Built-in Types' since Decimal is now implemented.

2. **Browser runtime precision**: Added a dedicated 'Browser runtime precision' note to §9.7.2 documenting that the Python runtime is exact while the browser MVP uses IEEE 754. Lists the affected functions and notes the planned fix (arbitrary-precision library).

3. **Resolver rejects Decimal\<T\>**: The Decimal branch now emits an E130 diagnostic if type arguments are provided. Added `test_decimal_rejects_type_args` to verify.

4. **Gating builtins for browser targets**: Respectfully disagree with this approach — the builtins work correctly on both targets, the browser just has reduced precision (documented in the spec). Gating them would make programs non-portable between targets. The precision limitation is a browser runtime implementation detail, not a language-level concern. A future browser runtime version using an arbitrary-precision decimal library will resolve this without any language changes.
